src/table/src/predicate.rs

```diff
@@ -66,6 +70,181 @@ impl Predicate {
     }
 }
 
+// tests for `TimeRangePredicateBuilder`locates in src/query/tests/time_range_filter_test.rs
+// since it requires query engine to convert sql to filters.
+pub struct TimeRangePredicateBuilder<'a> {
+    ts_col_name: &'a str,
+    filters: &'a [Expr],
+}
+
+impl<'a> TimeRangePredicateBuilder<'a> {
+    pub fn new(ts_col_name: &'a str, filters: &'a [Expr]) -> Self {
+        Self {
+            ts_col_name,
+            filters,
+        }
+    }
+
+    pub fn build(&self) -> TimestampRange {
+        let mut res = TimestampRange::min_to_max();
+        for expr in self.filters {
+            let range = self.extract_time_range_from_expr(expr.df_expr());
+            res = res.and(&range);
+        }
+        res
+    }
+
+    fn extract_time_range_from_expr(&self, expr: &DfExpr) -> GenericRange<Timestamp> {
+        match expr {
+            DfExpr::BinaryExpr(BinaryExpr { left, op, right }) => {
+                self.extract_from_binary_expr(left, op, right)
+            }
+            DfExpr::Between(Between {
+                expr,
+                negated,
+                low,
+                high,
+            }) => self.extract_from_between_expr(expr, negated, low, high),
+            DfExpr::InList {
+                expr,
+                list,
+                negated,
+            } => self.extract_from_in_list_expr(expr, *negated, list),
+            _ => TimestampRange::min_to_max(),
+        }
+    }
+
+    fn extract_from_binary_expr(
+        &self,
+        left: &DfExpr,
+        op: &Operator,
+        right: &DfExpr,
+    ) -> GenericRange<Timestamp> {
+        match op {
+            Operator::Eq => self
+                .get_timestamp_filter(left, right)
+                .map(TimestampRange::single)
+                .unwrap_or_else(TimestampRange::min_to_max),
+            Operator::Lt => self
+                .get_timestamp_filter(left, right)
+                .map(|ts| TimestampRange::until(ts, false))
+                .unwrap_or_else(TimestampRange::min_to_max),
+            Operator::LtEq => self
+                .get_timestamp_filter(left, right)
+                .map(|ts| TimestampRange::until(ts, true))
+                .unwrap_or_else(TimestampRange::min_to_max),
+            Operator::Gt => self
+                .get_timestamp_filter(left, right)
+                .map(TimestampRange::from)
+                .unwrap_or_else(TimestampRange::min_to_max),
+            Operator::GtEq => self
+                .get_timestamp_filter(left, right)
+                .map(TimestampRange::from)
+                .unwrap_or_else(TimestampRange::min_to_max),
```

Comment:
Do we need to add an inclusive option?

Reply:
Left-close right-close range cannot have exclusive start that generated by `Gt` operator, so `Gt` and `GtEq` are converted to `[start - 1, INF)` and `[start, INF)` respectively.

Reply:
It seems that both `Gt` and `GtEq` are converted into `[start, INF)`. The from_start method doesn't have options to control whether we can exclude the `start` like `until_end`.

Reply:
It seems that both `Gt` and `GtEq` are converted into `[start, INF)`. The from_start method doesn't have options to control whether we can exclude the `start` like `until_end`.

Reply:
> It seems that both `Gt` and `GtEq` are converted into `[start, INF)`. The from_start method doesn't have options to control whether we can exclude the `start` like `until_end`.

I made a mistake in previous comments. Conerting both`Gt` and `GtEq` to `[start, INF)` is valid, and there's no need to convert `Gt` to `[start-1, INF)`.